### ISSUE TITLE
Docker - Permission denied @ dir_s_mkdir Error for 1000 User

### ISSUE THREAD

**Issue by orangewolf** [scientist-softserv/louisville-hyku#66]
I'm getting a Permission denied @ dir_s_mkdir error in Docker when Hyku tries to write to file/directory. The owner of hyrax-webapp is 1000 rather than app.
Confirmed this is currently occurring on Ubuntu v20.04 running Docker v20.10.17.
I found the same problem addressed by Rob in 2020 on [Slack](https://samvera.slack.com/archives/C7E4KK8ER/p1592423006019500).

**Comment by orangewolf** [comment]
Shana July 2022
---
Hi [@rtkuehn](https://github.com/rtkuehn)
Could you copy/paste the exact error? Which directory is it?
In case we need it, could you send us some available days/times to meet so we can troubleshoot your issue together?
In the meantime, I'm also trying to hunt down the "permanent fix" solution to the conversation you posted above. Seeing that I don't have this issue, it'll be more challenging without access to the problem.

**Comment by orangewolf** [comment]
Shana July 2022
---
Noting he mentioned the short fix is to do the following. 

`docker-compose exec web chown -R app:app app/samvera/hyrax-webapp/tmp/cache`
 (<-- or to whatever directory it is)

**Comment by orangewolf** [comment]
Shana July 2022
---
@rtkuehn Chris Colvard [linked](https://github.com/samvera/hyku/pull/1693/files#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R168-R182) me to what he thinks was the fix at the time.

If the chown command from above works for you locally, you could try adding it to the docker-compose.yml command block under the web section.

Please give it a go and let us know.[@rtkuehn](https://github.com/rtkuehn) Chris Colvard [linked](https://github.com/samvera/hyku/pull/1693/files#diff-e45e45baeda1c1e73482975a664062aa56f20c03dd9d64a827aba57775bed0d3R168-R182) me to what he thinks was the fix at the time.
If the chown command from above works for you locally, you could try adding it to the docker-compose.yml command block under the web section.
Please give it a go and let us know.

**Comment by orangewolf** [comment]
Randy July 2022
---
@ShanaLMoore I'm seeing the error in the UI when attempting to add a file


![Image](https://user-images.githubusercontent.com/1054448/203825950-9bc49a4e-40c1-4fe5-80d5-0fd3134d8411.png)


The primary problem is that /app/samvera/hyrax-webapp and its files/folders (other than node_modules) is assigned uid=1000 & gid=1000. 



![Image](https://user-images.githubusercontent.com/1054448/203825993-537e73a8-91d2-461a-8d03-7734800524e2.png)



The docker file specifies that it should be 1001:101, but that didn't seem to happen. 


![Image](https://user-images.githubusercontent.com/1054448/203826073-cd1fad66-e0e1-4adc-8338-4573d424be43.png)



From what I understand, you won't be able to reproduce this on a Mac, which I reckon is why I'm first seeing it now that we're setting the Docker image up on Ubuntu.

The chown workaround isn't working for me either.

Outside container:
![Screen_Shot_2022-07-22_at_12.18.45_PM](/uploads/9426e9c6d70e4506caf572ba3a169266/Screen_Shot_2022-07-22_at_12.18.45_PM.png)

Inside docker-compose.yml as 'command: "chown -R app:app /app/samvera/hyrax-webapp"'


![Image](https://user-images.githubusercontent.com/1054448/203826155-bd65b463-e2b6-410d-9d90-6909e0577e7f.png)


I'm free 7/26 & 7/27 from 10am-2pm ET.

**Comment by orangewolf** [comment]
Shana July 2022
---
@rtkuehn 

When you cd into samvera/hyrax-webapp, what does ls -l look like after running this: 

`chmod -R 777 ./`

**Comment by orangewolf** [comment]
Randy July 2022
---
I get operation not permitted.

**Comment by orangewolf** [comment]
Randy July 2022
---
@ShanaLMoore I'll try Chris's solution and get back to you[@ShanaLMoore](https://github.com/ShanaLMoore) I'll try Chris's solution and get back to you

**Comment by orangewolf** [comment]
Shana July 2022
---
Also, what if you try it with sudo (if you are not root)? ref: https://askubuntu.com/questions/54525/chown-changing-ownership-of-operation-not-permitted

**Comment by orangewolf** [comment]
Randy July 2022
---
@ShanaLMoore 
I had the same result using sudo.

I also swapped the current docker-compose.yml code
```yaml
    command:
      - >
        chown -R app:app /app/samvera/hyrax-webapp/public/uploads &&
        chown -R app:app /app/samvera/hyrax-webapp/public/assets &&
        chown -R app:app /app/samvera/hyrax-webapp/tmp/cache
```
with Chris's syntax
```
    command:
      - >
        # Do not recurse through all of tmp. derivitives will make booting
        # very slow and eventually just time out as data grows
        chown -f app:app /app/samvera/hyrax-webapp/tmp
        chown -fR app:app /app/samvera/hyrax-webapp/tmp/pids
        chown -fR app:app /app/samvera/hyrax-webapp/tmp/uploads
        chown -fR app:app /app/samvera/hyrax-webapp/public/uploads
        chown -fR app:app /app/samvera/hyrax-webapp/public/assets
        chown -fR app:app /app/samvera/hyrax-webapp/tmp/cache
```
No luck :disappointed:

**Comment by orangewolf** [comment]
Shana July 2022
---
To confirm, did you try the sudo command outside of your container? @rtkuehnTo confirm, did you try the sudo command outside of your container? [@rtkuehn](https://github.com/rtkuehn)

**Comment by orangewolf** [comment]
Randy July 2022
---
Yep, I tried it outside the container.

**Comment by orangewolf** [comment]
Alisha July 2022
---
this may be related to the OS.

**Comment by orangewolf** [comment]
Randy July 2022
---
Based on what I found so far, this has occurred in various Linux distros and Windows. I'm still not sure why it's happening or how to fix it prior to build. I confirmed that it's happening with our code and the current Hyku release on Ubuntu 20. 
However...I was able to log into hyku_web_1 as root
(`docker exec -it --user root < container_id > bash`)
and successfully chown (`chown -R app:app /app/samvera/hyrax-webapp`) on a test linux box. I'm not sure if this is a permanent fix and I still need to confirm it & image uploads work on our dev machine.

**Comment by orangewolf** [comment]
Randy July 2022
---
@ShanaLMoore I can now confirm that the permissions are being modified via docker-compose.yml, but it seems maybe not all the necessary permissions.
Permissions are granted to tmp/cache in the docker-compose.yml. The original error I ran into was when hyku was trying to access the tmp dir. 


![Image](https://user-images.githubusercontent.com/1054448/203827923-471589a5-e581-47e1-89e5-54a64b4c85ff.png)



After granting permissions to tmp for the app user, I'm able to add an image in the UI without an error. However, I receive the following error on the show page.


![Image](https://user-images.githubusercontent.com/1054448/203827960-89a0ccf9-a5ab-45de-bc1c-5ca2453cc2c5.png)



I'm not sure why we're getting this error. Any ideas?

**Comment by orangewolf** [comment]
Shana July 2022
---
At first glance it is not clear to me why it's throwing an error. Thinking...

**Comment by orangewolf** [comment]
Randy July 2022
---
@ShanaLMoore @alishaevn It looks like our repos aren't synced. The error above is referencing a file we have (from: Revert "122-remove-auto-uv-search") but is no longer in your repo. How can we get back in sync?[@ShanaLMoore](https://github.com/ShanaLMoore) [@alishaevn](https://github.com/alishaevn) It looks like our repos aren't synced.
The error above is referencing a file we have (from: Revert "122-remove-auto-uv-search") but is no longer in your repo. How can we get back in sync?

**Comment by orangewolf** [comment]
Alisha July 2022
---
@rtkuehn looks like we removed the "\_universal_viewer.html.erb" file as part of the work for #122. it was submitted to your repo in https://github.com/rkuehn-uofl/hyku/pull/6, but you reverted that change in https://github.com/rkuehn-uofl/hyku/pull/7. that's why you have that file and we don't.[@rtkuehn](https://github.com/rtkuehn) looks like we removed the "_universal_viewer.html.erb" file as part of the work for [#122 (closed)](https://github.com/notch8/louisville-hyku/-/issues/122). it was submitted to your repo in https://github.com/rkuehn-uofl/hyku/pull/6, but you reverted that change in https://github.com/rkuehn-uofl/hyku/pull/7. that's why you have that file and we don't.

**Comment by orangewolf** [comment]
Randy July 2022
---
I think I may have reverted it because the merge erased some changes we made to the predicates.

**Comment by orangewolf** [comment]
Alisha July 2022
---
what do you mean by predicates?

**Comment by orangewolf** [comment]
Randy July 2022
---
It's the [metadata mapping](https://samvera.github.io/customize-metadata-model.html) in hyrax. Some of our metadata has changed a bit over time.It's the [metadata mapping](https://samvera.github.io/customize-metadata-model.html) in hyrax. Some of our metadata has changed a bit over time.

**Comment by orangewolf** [comment]
Alisha July 2022
---
ok. looking at the diffs for pr 6, nothing should have changed those. but you can edit/remove the files 1 by 1 if you feel something is related and interferes with your local changes.

**Comment by orangewolf** [comment]
Randy July 2022
---
I changed the ownership of hyrax-webapp/tmp in [docker-compose.yml](https://github.com/rkuehn-uofl/hyku/blob/cfacf2ca07ffcb67f06adf409f4247b92aa8282d/docker-compose.yml#L209) to fix the upload error we were seeing. The uv error was corrected by removing that \_universal_viewer.html.erb file. I don't know if any other permission related issues will pop up, but y'all can close this if you want.I changed the ownership of hyrax-webapp/tmp in [docker-compose.yml](https://github.com/rkuehn-uofl/hyku/blob/cfacf2ca07ffcb67f06adf409f4247b92aa8282d/docker-compose.yml#L209) to fix the upload error we were seeing. The uv error was corrected by removing that _universal_viewer.html.erb file. I don't know if any other permission related issues will pop up, but y'all can close this if you want.
